conda/core/envs_manager.py

```diff
@@ -299,155 +154,42 @@ def unregister_env(self, location):
-    #     pep = first(self._preferred_env_packages, lambda p: p['package_name'] == package_name)
-    #     return pep and pep['preferred_env_name']
-    #
-    # def get_registered_packages(self):
-    #     # returns Map[package_name, env_name]
-    #     return {pep['package_name']: pep for pep in self._preferred_env_packages}
-    #
-    # def get_registered_packages_keyed_on_env_name(self):
-    #     get_env_name = lambda x: x['preferred_env_name']
-    #     return groupby(get_env_name, self._preferred_env_packages)
-    #
-    # def get_private_env_prefix(self, spec_str):
-    #     package_name = spec_str.split()[0]
-    #     pep = first(self._preferred_env_packages, lambda p: p['package_name'] == package_name)
-    #     return join(self.envs_dir, pep['preferred_env_name']) if pep else None
-    #
-    # def get_preferred_env_package_entry(self, spec_str):
-    #     package_name = spec_str.split()[0]
-    #     pep = first(self._preferred_env_packages, lambda p: p['package_name'] == package_name)
-    #     return pep or None
-
-
-def determine_target_prefix(ctx, args=None):
-    """Get the prefix to operate in
-
-    Args:
-        ctx: the context of conda
-        args: the argparse args from the command line
-
-    Returns: the prefix
-    Raises: CondaEnvironmentNotFoundError if the prefix is invalid
-    """
+    def _clean_environments_txt(self, remove_location=None):
+        if not isfile(self.catalog_file):
+            return ()
 
-    argparse_args = args or ctx._argparse_args
-    try:
-        prefix_name = argparse_args.name
-    except AttributeError:
-        prefix_name = None
-    try:
-        prefix_path = argparse_args.prefix
-    except AttributeError:
-        prefix_path = None
-
-    if prefix_name is None and prefix_path is None:
-        return ctx.default_prefix
-    elif prefix_path is not None:
-        return expand(prefix_path)
-    else:
-        if '/' in prefix_name:
-            raise CondaValueError("'/' not allowed in environment name: %s" %
-                                  prefix_name)
-        if prefix_name in (ROOT_ENV_NAME, 'root'):
-            return ctx.root_prefix
-        else:
-            try:
-                return EnvsDirectory.locate_prefix_by_name(prefix_name)
-            except EnvironmentNameNotFound:
-                return join(EnvsDirectory.first_writable().envs_dir, prefix_name)
+        environments_txt_lines = list(yield_lines(self.catalog_file))
+        try:
+            location = normpath(remove_location or '')
+            idx = environments_txt_lines.index(location)
+            del environments_txt_lines[idx]
+        except ValueError:
+            # remove_location was not in list.  No problem, just move on.
+            pass
+
+        real_prefixes = tuple(p for p in environments_txt_lines if self.is_conda_environment(p))
+        if self.is_writable:
+            with open(self.catalog_file, 'w') as fh:
+                fh.write('\n'.join(real_prefixes))
```

Comment:
A new line is missing here. This should be
```
fh.write('\n'.join(real_prefixes) + '\n')
```

Reply:
If the new line is missing at the end, then the next prefix will get appended on the previous line, causing something like this:
```
/tmp/test1/tmp/test2
```
And a consecutive call to `conda env list` will remove this line itself, since it might not be a valid folder

Reply:
Perhaps a test case could cover this scenario too?